### ISSUE TITLE
Errata Issue O - MPI_WIN_TEST with same progress as MPI_TEST

### ISSUE THREAD

**Issue by RolfRabenseifner** [mpi-forum/mpi-issues#499]
# Problem

MPI-3.1 does not define that MPI_WIN_TEST must do progress, i.e., does not define that the same rule applies as for pt-2to-pt MPI_TEST

# Proposal

Add
> The progress rule for MPI_TEST, as described in Section 3.7.4, also applies to MPI_WIN_TEST.

(same sentence as for MPI_REQUEST_GET_STATUS)

# Changes to the Text

The new additional sentence for MPI_WIN_TEST is
> The progress rule for MPI_TEST, as described in Section 3.7.4, also applies to MPI_WIN_TEST.

# Impact on Implementations

None, because existing MPI applications and the used MPI libraries already use respectively implemented progress for MPI_WIN_TEST.

# Impact on Users

Clarity that existing usage is correct.

# References and Pull Requests

(PR https://github.com/mpi-forum/mpi-standard/pull/591 was MPI-4.0)
PR https://github.com/mpi-forum/mpi-standard/pull/729 for MPI-4.1

**Comment by RolfRabenseifner** [comment]
I assigned it to the chairs and to members of the three chapter committees: 
Terms (progress), 1-sided (MPI_Win_TEST), and pt-to-pt (MPI_TEST, MPI_I(M)PROBE, and MPI_REQUEST_GET_STATUS)

**Comment by jeffhammond** [comment]
I assume the text intended as the target of this reference is:

> If an MPI_TEST that completes a receive is repeatedly called with the same arguments,
and a matching send has been started, then the call will eventually return flag = true, unless
the send is satisfied by another receive. If an MPI_TEST that completes a send is repeatedly
called with the same arguments, and a matching receive has been started, then the call will
eventually return flag = true, unless the receive is satisfied by another send.

This text is not applicable to `MPI_Win_test` because `MPI_Win_test` does not complete sends or receives, it completes an RMA exposure epoch.

If you intended to require `MPI_Win_test` to make progress on send-recv, then I disagree with that goal.  I do not want to burden RMA implementations with the need to poke message queues.

**Comment by RolfRabenseifner** [comment]
In PR591, after a longer discussion, we agreed on the following sentence with **eventually**:

>   Repeated calls to \mpifunc{MPI\_WIN\_TEST} with the same \mpiarg{win} argument
   will eventually return \mpiarg{flag}\code{ = true} once
   all accesses to the local window by the group to which it was exposed by
   the corresponding \mpifunc{MPI\_WIN\_POST} call have been completed
   as indicated by matching \mpifunc{MPI\_WIN\_COMPLETE} calls,
   i.e., once a call to \mpifunc{MPI\_WIN\_WAIT} instead of \mpifunc{MPI\_WIN\_TEST}
   would behave as if it were local.

but this is a real change, not an addition, because there exists already
a clear wording in the definition of MPI_Win_test:

>   It returns \mpiarg{flag}\mpicode{ = true} if all accesses
   to the local window by the group to which it was exposed by
   the corresponding \mpifunc{MPI\_WIN\_POST} call have
   been completed as indicated by matching
   \mpifunc{MPI\_WIN\_COMPLETE} calls, and
   \mpiarg{flag}\mpicode{ = false} otherwise.
   In the former case \mpifunc{MPI\_WIN\_WAIT} would behave
   as if it were local.

Martin Schulz detected in
https://github.com/mpi-forum/mpi-standard/pull/591/files#r617251572
that with this current wording, the following program

>commA and commB should have same ranks for processes rank 0 and 1
winA is a window handle based on commA
if(myrank==0) { // target
` `  MPI_Win_post(group_of_commA_rank_1_only, /*assert*/ 0, winA);
` `  MPI_Recv(/**buf*/..., /*source*/ 1, ..., commB, ...);
` `  MPI_Win_test(winA, &flag);
` `  if (!flag) {
` ` ` `   printf("Broken MPI library according to definition of MPI_Win_test in MPI-3.0 and 4.0\n");
` ` ` `   MPI_Win_wait(winA);
` `  }
} else if (myrank==1) { // origin
` `  MPI_Win_start(group_of_commA_rank_0_only, /*assert*/ 0, winA);
` `  MPI_Win_complete(winA); // has somehow to sent a flag to rank 0 in winA=commA
` `  MPI_Send(/*token*/..., /*dest*/ 0, ..., commB);
}

should in principle never report a broken MPI library.

If commA == commB, then this would mean that the MPI_Send in commB must
not be transmitted faster than internal message of MPI_Win_complete.
Such non-overtaking is defined in MPI-3.5 Sect 3.5 Paragraph "Order"
only for point-to-point communication.

If we do the thought experiment that commA and commB are produced through
MPI_Connect with commB using a transmission method for the pt-to-pt message
with a 100 times shorter latency than in commA for the RMA synchronization,
then I would expect that all MPI implementations would print
"Broken MPI library according to definition of MPI_Win_test in MPI-3.0 and 4.0",
i.e., are somehow broken according to the current wording,
or in other words, only the new wording would be a correct description of the
current MPI libraries.

**Summary:**

The current wording in MPI-3.0 page 444 lines 33-36 (already with a small coorction in MPI-4.0)

>   It returns \mpiarg{flag}\mpicode{ = true} if all accesses ... (text see above)

should be substituted by the new wording

>   Repeated calls to \mpifunc{MPI\_WIN\_TEST} with the same \mpiarg{win} argument ... (text see above)

**Comment by raffenet** [comment]
> If commA == commB, then this would mean that the MPI_Send in commB must
not be transmitted faster than internal message of MPI_Win_complete.
Such non-overtaking is defined in MPI-3.5 Sect 3.5 Paragraph "Order"
only for point-to-point communication.

Why would the non-overtaking rule apply to internal library communication?

**Comment by bosilca** [comment]
The text in MPI_Win_test is much more clear than the proposed addition. Second, why do you expect a sequential ordering between pt-to-pt and RMA operations on different MPI objects ?

**Comment by Wee-Free-Scot** [comment]
Unless certain patterns of MPI calls are specifically called out and prohibited (they are not, AFAIK), then the following is a valid MPI program that requires repeated calls to MPI_WIN_TEST to guarantee implicitly mandated progress of a point-to-point operation as part of the explicitly mandated guarantee of progress for the one-sided synchronisation operation its semantics describe.

``` c
MPI_INIT
if (0==myrank) {
  MPI_WIN_CREATE(myAckSize)
  MPI_WIN_POST(win)
  MPI_ATTACH(myHugeBuffer)
  MPI_BSEND(myHugeMessage)
  while(!flag) MPI_WIN_TEST(&flag, win)
  MPI_DETACH(myHugeBuffer)
} else if (1==myrank) {
  MPI_WIN_CREATE(zeroSize)
  MPI_RECV(myHugeMessage)
  MPI_WIN_START(win)
  MPI_PUT(myAck)
  MPI_WIN_COMPLETE(win)
}
MPI_FINALIZE
```

The buffered mode send is defined by MPI to be local — it is permitted to delay its return until it has buffered the huge message in the attached huge buffer (locally) but it must not delay its return until the huge message is delivered (remotely) because that causes its return to depend on some specific remote MPI procedure (the matching receive procedure).

Thus, the loop containing MPI_WIN_TEST can be entered before the message data has been transmitted but that data must be transmitted before the matching receive procedure is permitted to return at the other MPI process. Every call to the MPI_WIN_TEST procedure will return flag==false until the other MPI process calls MPI_WIN_COMPLETE, which it will not do until the receive operation is complete and the MPI_RECV procedure returns, which cannot happen unless it receives the huge message, which cannot happen unless the huge message is transmitted by the process with rank 0 in MPI_COMM_WORLD, which has buffered the huge message locally but is now busy waiting in a tight loop over MPI_WIN_TEST.

If repeated calls to MPI_WIN_TEST is required to guarantee progress of all outstanding non-local operations in which it is involved, then all is well. If not, then this program could deadlock with certain MPI implementations, e.g. one that chooses to postpone progress until it is absolutely required by the MPI Standard (in this program the attached buffer must be emptied no later than the return of MPI_BUFFER_DETACH, which will delay its return until all messages currently in the buffer have been transmitted).

This is a design decision — should the above program (and all similar programs in an ill-defined class of programs that, in some sense, “suffer the same problem”) be summarily declared to be erroneous? They are not, currently.

**Comment by RolfRabenseifner** [comment]
My question to @bosilca and @raffenet: How do you interpret the current words

> It returns \mpiarg{flag}\mpicode{ = true} if all accesses to the local window by the group to which it was exposed by
the corresponding \mpifunc{MPI_WIN_POST} call have been completed as indicated by matching \mpifunc{MPI_WIN_COMPLETE} calls, and \mpiarg{flag}\mpicode{ = false} otherwise. 

In my example, I guarantee that MPI_Win_test is called after MPI_Win_complete is called by all processes that called corresponding MPI_Win_post.
How should an MPI library guarantee that it really returns flag=true in such case?

@Wee-Free-Scot : Very interesting example because it shows that the old wording and perhaps the new wording may be not enough to guarantee this. The current wording for MPI_Test and the proposed wording of MPI_Win_test seem not to be strong enough to guarantee progress in other chapters of MPI.

**Comment by raffenet** [comment]
> My question to @bosilca and @raffenet: How do you interpret the current words
> 
> > It returns \mpiarg{flag}\mpicode{ = true} if all accesses to the local window by the group to which it was exposed by
> > the corresponding \mpifunc{MPI_WIN_POST} call have been completed as indicated by matching \mpifunc{MPI_WIN_COMPLETE} calls, and \mpiarg{flag}\mpicode{ = false} otherwise.
> 
> In my example, I guarantee that MPI_Win_test is called after MPI_Win_complete is called by all processes that called corresponding MPI_Win_post.
> How should an MPI library guarantee that it really returns flag=true in such case?

The `MPI_WIN_COMPLETE` may indeed use the same communication resources and ordering as the send/recv in your example, but it also might not. I don't see how the standard requires ordering in this case. IMO, there is no guarantee that the implementation has observed the `MPI_WIN_COMPLETE` on rank 0 at the time of `MPI_WIN_TEST` in your example.

**Comment by bosilca** [comment]
My understanding is that @RolfRabenseifner example is about communication ordering while @Wee-Free-Scot is looking at progress guarantee. Too many critical topics going on in parallel on this issue to be able to have a meaningful discussion. 

But, I have to join @raffenet, because @RolfRabenseifner example __cannot__ guarantee that `MPI_WIN_TEST` is called after the impact of all other processes having called the corresponding `MPI_WIN_COMPLETE` is visible at the target, because, as @raffenet highlighted, there is no ordering at the network level between communications for RMA and for pt2pt. Your example, is based on the assumption that the sequential program order is respected by network events, a heresy even if the windows was created using the same communicators as for the pt2pt communications. If we change the standard in order to enforce such an ordering it will have a drastic impact on performance.

Moreover, the non-overtaking assertion is about the matching of pt2pt events and not about their completion. There is nothing in the MPI standard that defines the order in which pt2pt communications (and therefore their associated requests) must complete.

Regarding the progress guarantee, sure we can standardize but we need to be aware of the implications. From an implementation perspective, this might impose undesirable overheads in many cases, because progress must now be global. But more concerning is that we will need to address not only pt2pt and RMA, but all types of MPI related data movement, including files and generalized requests, which in few MPI implementations had lower frequency of query.

**Comment by devreal** [comment]
Regarding progress:
From a user's perspective, I would expect the example of @Wee-Free-Scot to be a correct MPI application, based on several statements in the standard, including in §3.7.4 on P2P:


> A call to MPI_WAIT that completes a receive will eventually terminate and return if a matching send has been started [...]. In particular, if the matching send is nonblocking, then the receive should complete even if no call is executed by the sender to complete the send.

Interestingly, the verbiage for MPI_Test does not contain the second part of this quote. As a user I would assume that the same rules apply though.

Also, example 6.31 on non-blocking collectives and P2P

> The MPI library must progress the barrier in the MPI_Recv call.

And the Rationale in §12.7.3:

> MPI implementations must guarantee that a process makes progress on all enabled communications it participates in, while blocked on an MPI call.

I notice that the two statements reference blocking procedure calls but I don't see why this would not apply to non-blocking procedure calls.

The Rationale in §12.7.3 also states:
> According to another interpretation, the complete call may block until process 1 reaches the wait call, or reaches another MPI call.

(this is the *weaker* interpretation of the RMA progress rules)

This suggests that *any* MPI call is sufficient to progress communication.

@bosilca and @jeffhammond seem to argue that the example provided by @Wee-Free-Scot is not a correct MPI application. Different people interpreting the same text differently suggests that this needs clarification in the standard. Currently, the progress rules are scattered across the document with unclear language (due to the incremental growth of the standard). A more centralized approach would be helpful (clearly stating what guarantees user can and cannot expect).

Regarding ordering:
I agree with @bosilca that the ordering @RolfRabenseifner wants to enforce here might incur additional overheads and the benefits are dubious. The guarantee that the call to `MPI_Win_test` returns `flag = 1` immediately in this example is of little help in reality.

**Comment by RolfRabenseifner** [comment]
First @devreal : Yes, there are these text locations on blocking calls and one of the few progress rules is that any blocking call somehow guarantee progress on any internally ongoing operation.
The question that I address here is about the TEST calls. They are not blocking. And there was in MPI-3.1 the "eventually" statement for MPI_Test. This statement guarantees that progress must happen if an application calls MPI_Test in a loop.
It does not say MPI_Test always must make progress; for example, a background thread may make this progress.
With Issue #468 we expand it to MPI_Request_get_status.
And in this issue, we try to do this also for MPI_Win_test.

Next @bosilca and @raffenet : To be very clear: I do not want that MPI is internally slowed down by some message ordering.
I want to show, that the current wording is not correct. It says

> It returns flag=true if all accesses to the local window by the group to which it was exposed by the corresponding MPI_WIN_POST call have been completed as indicated by matching MPI_WIN_COMPLETE calls, and flag=false otherwise.

In my example (it has only two processes, one origin and one target process), it is guaranteed that MPI_WIN_COMPLETE is called by all origins and therefore the current wording does allow only that flag=true is returned. 
I fully agree when you or somebody states that 
- current MPI implementations cannot guarantee this, and
- that nobody wants to have such guarantee because it would slow down the MPI calls.

I hope that you also agree that therefore the current wording should be substituted.
Q1: Do you agree?

The second discussion is about how to substitute the wording. 
And here, @Wee-Free-Scot 's example shows that the propsed new text may still not be enough.

But before thinking on wording, we should agree on that @Wee-Free-Scot 's example is a correct MPI program.
Q2: Do you agree?

**Comment by Wee-Free-Scot** [comment]
Careful now -- both @devreal "the two statements reference blocking procedure calls" and @RolfRabenseifner "any blocking call" change from what is stated in the MPI Standard "while blocked on an MPI call" into something else by replacing a word ("blocked") with a term ("blocking").

It is possible to be blocked on a nonblocking MPI call, just as it is possible to not need to block on a blocking MPI call. This is because blocking and nonblocking are MPI terms with a specific meaning that has very little to do with the verb "to block" in any of its forms.

IMHO "while blocked on an MPI call" roughly means "while prevented from returning because it cannot yet meet its semantic specification" and not "while executing a call of a blocking procedure".

Thus, @devreal: yes, the rationale can apply to nonblocking procedures as well as blocking procedures.

Also, @RolfRabenseifner: we should probably say either "MPI_Test is nonblocking" (which is true but seems irrelevant here) or "MPI_Test does not block", i.e. it is never prevented from returning due to an inability to meet its semantic specification.

**Comment by Wee-Free-Scot** [comment]
One point about "it is guaranteed that MPI_WIN_COMPLETE is called by all origins and therefore the current wording does allow only that flag=true is returned" if I may.

IMHO, we must be acknowledge a non-zero time delay for information to propagate from one process to another (at least in part because physics) -- in practice, that means despite the fact that "MPI_WIN_COMPLETE is called by all origins" (a past event), it may also be simultaneous true that not all targets are yet aware of this. This is the key reason for the "eventually" occurring in the statements about the subsequent consequences. Eventually (it is presumed) through the natural course of events (plus or minus progress, which might require at least one MPI call) all targets will become aware of what the origin processes already know. At that point in time, some arbitrary interval of time after the event actually happened, it will become possible for MPI_WIN_TEST to return flag==true.

However, I see no way for a user to distinguish between a slow/late delivery of an internal protocol "message" and an implementation that chooses to insert an arbitrary delay after seeing such a protocol message before it deigns to pass this information on to the user. An implementation that chose to return flag==false for the next 6 calls after all protocol messages arrive that permit it to return flag==true, would be strange indeed but valid, IMHO. In the same way, returning flag==false for all calls within 5 seconds of the last needed protocol message, would also be strange but valid.

Unless we create a definition of a QoS -- a maximum time from the last needed MP_WIN_COMPLETE call until the target must indicate it has got the resulting protocol message by returning flag==true -- then we must allow an arbitrary delay. An arbitrary delay for one reason is indistinguishable from an arbitrary delay for some other reason.

Just to be clear, the QoS suggestion is a straw person, and an unimplementable/terrible idea, IMHO. Also, the implementation(s) that deliberately insert unnecessary delays are straw person examples.

This leads me towards the conclusion that the example program given by @RolfRabenseifner is unsafe -- it might work sometimes, but there is no guarantee that it will work, i.e. there is a non-zero probability that it will report a broken MPI library for an MPI library that is not, in fact, broken.

**Comment by wgropp** [comment]
@Wee-Free-Scot makes an important point on the QoS and the issues of what is required. We had similar issues with Fault Tolerance in the case of unresponsive processes - it is unknowable whether the process is really dead or just very, very slow in responding. Adding strong requirements may lead to letter-of-the-law implementations, such as one major vendor that used a 500+ millisecond timer interrupt to schedule a progress thread, that don't address the user's expectations for performance.

What users need to know is how to write code that will work, and what special steps they may need to take to make the code fast. Skilled users will avoid corners in the specification where design tradeoffs in implementations may cause problems. Guaranteeing the code will execute, but perhaps slowly, or that *all* MPI implementations will be slow, is not in anyone's interest.

In MPI-2, we specified that some RMA operations were erroneous, particularly with overlapping access.  This seemed reasonable at the time. But for some users, and other interface designs, such as using MPI RMA for UPC, this added requirement meant that MPI RMA could not be easily used for some natural cases. In MPI-3, we made a semantic change that these operations were no longer erroneous - rather, they were undefined. That gave implementations (and some algorithms) more flexibility to provide what users are seeking when they go to parallelism - performance.

**Comment by Wee-Free-Scot** [comment]
With the bug-hunting help of @jsquyres, I have written this program, which supports (but does not prove) my pseudo-code example:
``` c
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char **argv) {
  int myRank=-1, tag=101, myAck=222, ranks[1], flag=0, size=1000*1000*1000, count=size/sizeof(double);
  void *pTemp;
  MPI_Win win;
  MPI_Group grpWorld, grpOther;
  double *myHugeMessage = malloc(size);
  char *myHugeBuffer = malloc(size+MPI_BSEND_OVERHEAD);

  printf("Got here 0\n");
  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &myRank);
  printf("[rank:%d] Got here 1\n", myRank);
  MPI_Comm_group(MPI_COMM_WORLD, &grpWorld);
  printf("[rank:%d] Got here 2\n", myRank);
  MPI_Win_create(&myAck, sizeof(int), sizeof(int), MPI_INFO_NULL, MPI_COMM_WORLD, &win);
  printf("[rank:%d] Got here 3\n", myRank);
  if (0==myRank) {
    ranks[0] = 1;
    MPI_Group_incl(grpWorld, 1, ranks, &grpOther);
  printf("[rank:%d] Got here 3.1\n", myRank);
    MPI_Win_post(grpOther, 0, win);
  printf("[rank:%d] Got here 3.2 pointer is %p\n", myRank, myHugeBuffer);
    MPI_Buffer_attach(&(myHugeBuffer[0]), size+MPI_BSEND_OVERHEAD);
  printf("[rank:%d] Got here 3.3 pointer is %p\n", myRank, myHugeBuffer);
    MPI_Bsend(&myHugeMessage[0], count, MPI_DOUBLE, 1, tag, MPI_COMM_WORLD);
  printf("[rank:%d] Got here 3.4\n", myRank);
  sleep(10);
  printf("[rank:%d] Got here 3.5\n", myRank);
    while(!flag) MPI_Win_test(win, &flag);
  printf("[rank:%d] Got here 3.6\n", myRank);
    MPI_Buffer_detach(&pTemp, &size);
  } else if (1==myRank) {
    //MPI_Win_create(&myAck, 0, sizeof(int), MPI_INFO_NULL, MPI_COMM_WORLD, &win);
    ranks[0] = 0;
    MPI_Group_incl(grpWorld, 1, ranks, &grpOther);
  printf("[rank:%d] Got here 3.1\n", myRank);
    MPI_Recv(&myHugeMessage[0], count, MPI_DOUBLE, 0, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  printf("[rank:%d] Got here 3.2\n", myRank);
    MPI_Win_start(grpOther, 0, win);
  printf("[rank:%d] Got here 3.3\n", myRank);
    MPI_Put(&myAck, 1, MPI_INT, 0, 0, 1, MPI_INT, win);
  printf("[rank:%d] Got here 3.4\n", myRank);
    MPI_Win_complete(win);
  }
  printf("[rank:%d] Got here 4\n", myRank);
  MPI_Win_free(&win);
  printf("[rank:%d] Got here 5\n", myRank);
  MPI_Finalize();
  printf("[rank:%d] Got here 6\n", myRank);
}
```

This program runs to completion successfully using Open MPI 4.0.5 on a MacBook Pro using `mpi exec -n 2 ./a.out` at the command line.

The timing and ordering of the output tells me that Open MPI buffers the send message locally and does not complete transmitting the message data to the other process until the loop over MPI_WIN_TEST, which must be guaranteeing progress of the point-to-point message. This is suggested by the other process getting to, but not returning from, the MPI_Recv call until the sending process enters the loop over MPI_WIN_TEST.

This shows me that at least one MPI implementation has chosen to delay progress of a point-to-point message until a one-sided synchronisation call requires it to guarantee that progress actually happens.

**Comment by Wee-Free-Scot** [comment]
This issue is important, but it is being superseded by an attempt to create a single holistic definition of progress in the semantic terms chapter. @RolfRabenseifner is there an issue for your efforts to define progress in Chapter 2?

**Comment by devreal** [comment]
I removed this issue from the 4.1 project, feel free to add it back if you want to work on it for 4.1.

**Comment by RolfRabenseifner** [comment]
Dear Joseph @devreal, it is important that the definitions of each test procedure (and MPI_Win_test is such a test procedure) are defined correctly. PR https://github.com/mpi-forum/mpi-standard/pull/591 had a good result (maybe not perfect, but perfect is always too expensive). Therefore, I ask you to 
- bring this issue back to MPI-4.1
- add a small PR that substitutes the text MPI-4.0 page 601 lines 28-31

> This is the nonblocking version of MPI_WIN_WAIT. It returns flag = true if all accesses to the local window by the group to which it was exposed by the corresponding MPI_WIN_POST call have been completed as signalled by matching MPI_WIN_COMPLETE calls, and flag = false otherwise. 

by

> \mpifunc{MPI\_WIN\_TEST} is local. Repeated calls to \mpifunc{MPI\_WIN\_TEST} with the same \mpiarg{win} argument will eventually return \mpiarg{flag}\code{ = true} once all accesses to the local window by the group to which it was exposed by the corresponding \mpifunc{MPI\_WIN\_POST} call have been completed as indicated by matching \mpifunc{MPI\_WIN\_COMPLETE} calls, and \mpiarg{flag}\code{ = false} otherwise.

It was already agreed that the current text is wrong, because an MPI library may need some time until all MPI_Win_complete are propagated. Therefore it must be substituted. The "eventually" implies that MPI_Win_test must do progress, without explicitely touching this topic, i.e., exactly the same as for MPI_Test.

You can do it as an errata, because the current text is definitely wrong.

The changelog sentence can simply be "The definition of MPI_WIN_TEST was clarified".
Technically, there is no problem, because no implementor tried to implement the current text whic required infinite speed of light ;-) I also expect that all implementors provide progress within MPI_Win_test.

**Comment by RolfRabenseifner** [comment]
@wesbland @devreal 
> I removed this issue from the 4.1 project, feel free to add it back if you want to work on it for 4.1.

Please can you put it back to the MPI-4.1 project.

**Comment by RolfRabenseifner** [comment]
PDF file: [mpi41-report_Issue499_PR729.pdf](https://github.com/mpi-forum/mpi-standard/files/10024194/mpi41-report_Issue499_PR729.pdf), see page pages 569 and 999

**Comment by wesbland** [comment]
This passed an [errata vote](https://mpi-forum.org/meetings/2022/12/votes) on 2022-12-07.

<table>
<tr>
	<td> Yes
	<td> No
	<td> Abstain
<tr>
	<td> 30
	<td> 0
	<td> 3
</table>

**Comment by wesbland** [comment]
This passed an [errata vote](https://mpi-forum.org/meetings/2022/12/votes) on 2022-12-07.

<table>
<tr>
	<td> Yes
	<td> No
	<td> Abstain
<tr>
	<td> 30
	<td> 0
	<td> 3
</table>
